src/Ninject.Web.AspNetCore/NinjectServiceProvider.cs

```diff
@@ -16,24 +17,36 @@ namespace Ninject.Web.AspNetCore
 	public class NinjectServiceProvider : IServiceProvider, ISupportRequiredService
 	{
 		private readonly IKernel _kernel;
-		
+
 		public NinjectServiceProvider(IKernel kernel)
 		{
 			_kernel = kernel;
 		}
 
 		public object GetRequiredService(Type serviceType)
 		{
-			var result = _kernel.Get(serviceType);
+			var result = _kernel.Get(serviceType, LatestBindingConstraint);
 			return result;
 		}
 
 		public object GetService(Type serviceType)
 		{
-			// call TryGet as IServiceProvider.GetService must return null if not found.
-			var result = _kernel.TryGet(serviceType);
-			return result;
+			// Call TryGet as IServiceProvider.GetService must return null if not found.
+			// See also https://github.com/ninject/Ninject/issues/378 and NinjectBugsRegressionTest
+			try
+			{
+				var result = _kernel.TryGet(serviceType, LatestBindingConstraint);
+				return result;
+			}
+			catch (InvalidOperationException)
+			{
+				return null;
+			}
 		}
 
+		private bool LatestBindingConstraint(IBindingMetadata bindingMetadata)
+		{
+			return bindingMetadata.Get<BindingIndex.Item>(nameof(BindingIndex))?.IsLatest ?? true;
```

Comment:
Clever idea 👍

Reply:
I wonder about one point. If we have a class bound with Ninject via Kernel.Bind and want to inject via constructor injection a dependency which has been added via the service provider way, it will fail because Ninject doesn't resolve with the metadata right? To make that work, woudln't we need to bind the latest one also without binding metadata? I also wonder about the constructor injection in ASP.NET Core itself. It will call GetService on the ServiceProvider for the root class, but the constructor parameters are created by the kernel directly right. So it doesn't has the metadata. 

Would be great to add also a testcase for constructor injection as well.